src/Routing/Router.php

```diff
@@ -488,12 +497,47 @@ public static function url($url = null, bool $full = false): string
 
         if (empty($url)) {
             $output = $context['_base'] . ($here ?? '/');
-            if ($full) {
+            if ($full === true) {
                 $output = static::fullBaseUrl() . $output;
             }
 
             return $output;
         }
+
+        if (is_string($url)) {
+            $plainString = (
+                strpos($url, 'javascript:') === 0 ||
+                strpos($url, 'mailto:') === 0 ||
+                strpos($url, 'tel:') === 0 ||
+                strpos($url, 'sms:') === 0 ||
+                strpos($url, '#') === 0 ||
+                strpos($url, '?') === 0 ||
+                strpos($url, '//') === 0 ||
+                strpos($url, '://') !== false
+            );
+
+            if ($plainString) {
+                return $url;
+            }
+
+            if ($url[0] === '@' && strpos($url, '::') > 1) {
```

Comment:
I am wondering if instead of using `::` as separator between controller and action names could be use `@` instead, like `Articles@view`?

By doing so we would just have to check if string has `@` character to determine if it's "path" string. We could use the `@` prefix instead to denote a named route. So the check would become like:

```php
$pos = strpos('@', $url);
// It's a named or path route
if ($pos !== false) {
    // It's a named route
    if ($pos === 1) {
    }
}
```

Reply:
The @ is a bit harder to type though on most Keyboards.

Reply:
`::` is already used by `connect()`, so I wouldn't like one more format. 

I would prefer different prefix for named routes.

Reply:
> The @ is a bit harder to type though on most Keyboards.

You already have to type @ as prefix in current proposal, so I don't see how your argument helps :) My proposal avoids having to type two additional colons.

> :: is already used by connect(), so I wouldn't like one more format.

Ah right, maintaining similarly when setting up routes and generating URLs makes sense.

But with the @ prefix is the additional overhead of strpos() call to check for 2 colons necessary?

Reply:
Isnt the @ optional and you can still use the _path key?
Also once is different than multiple times.

Reply:
> Isnt the @ optional and you can still use the _path key?

There's no `_path` key. The consensus was to directly pass string as 1st argument to `Router::url()` and using array with key is not convenient enough.

> Also once is different than multiple times.

I proposal was to remove `@` as 1st character and replace `::` with `@`, so it's still only once.

Reply:
Problem with string as 1st argument is how to use this from `HtmlHelper::link()`, `UrlHelper::build()` etc, as there is no way to pass additional params with string.

I'm thinking if I need to return `_path` as alternate so those methods could use array but that would one more way to do same...
```php
HtmlHelper::link('View', ['_path' => 'Articles::view', 3]);
HtmlHelper::link('View', Router::url('Articles::view', [3]));
```

 Any other ideas?

Reply:
@garas It seems we'll have to add back the `_path` key then. `['_path' => 'Articles::view', 3]` is still more convenient than having to use `Router::url('@Articles::view', [3])`.

Reply:
Would using a format like `['Articles::view', [3]]` work? The tuple of string with `::` in the 0th slot and an array in the 1st slot could help us detect short names with parameters.

Reply:
I like tuple format, more compact and no magic `_path` string.

Reply:
What if you omit 2nd argument? Only path argument, the first?

Reply:
Then string only `@Articles::index`, no array.

Reply:
If we can skip the `@` that would be ideal as we don't use `@` when connecting routes and only having one syntax would be :+1: It should be rare for an existing URL to match both the new short syntax and be a useful URL.

Instead of making the new short syntax complex to preserve compatibility with what should be a rare situation could we provide an escape hatch for this situation? Perhaps something like

```php
Router::url(['_path' => 'MyThing:photo'])
```

Which is treated like a bare string URL would have in the past?

Reply:
So which of these it should support? With or without `@` prefix?

1. Bare string. Cleanest, but extra params not available for `HtmlHelper::link()`. Safer with `@` prefix.
```php
Router::url('Articles::index');
Router::url('Articles::view', [3]);
```
2. Tuple format. Cleanest for functions like `HtmlHelper::link()`.
```php
Router::url(['Articles::index', []]); // empty extra params array required
Router::url(['Articles::view', [3]]);
```
3. Using special key. Safest. Very similar to `_name`.
```php
Router::url(['_path' => 'Articles::index']);
Router::url(['_path' => 'Articles::view', 3]);
```

All variants has pros and cons.

Reply:
1 with the prefix and 3 could both be valid and supported, right?
I don't think 2 is good due to the decreased usability.

Reply:
Ok let, me take the "path part only" back, looks like currently the "plain string" check looks for `?` and `#` at position `0`. But testing only the path part for `::` should be easy enough :)

Reply:
Combining 1 and 3 would let us handle 0 and many parameters and shouldn't overlap with existing signature forms. We'd need to add integration tests to each place where we generate URLs though to make sure with any approach we choose.

Reply:
I don't want anyone to blame me later on, so I had a further look at using `/` for relative URLs to avoid conflicts :upside_down_face:  It looks like it shouldn't be a problem for non asset URLs, as relative URLs processed by the router get a `/` prepended anyways.

For assets, `Router::url()` seems to be only used by the helper for URLs where a scheme is present, that is URLs containing either `://`, `a-z0-9://`, or `a-z:`. The latter currently poses a limitation to using `::`, it would currently only work when the part before the `::` would contain non-`a-z0-9` chars, otherwise it would be interpreted as a URI scheme, ie `foo/bar::baz` would work and for example map to `css/foo/bar::baz.css`, while `foo::bar` would not work, as `foo:` would be interpreted as a scheme.

Long story short, so far the only string URLs that I found that might not be easily upgradable to avoid possible conflicts, would be asset URLs like `foo/bar::baz` (one could of course do `/css/foo/bar::baz`, but that would mean loosing path prefixing), and ones that have a scheme but don't use `://`, for example `cid:my::asset::id`.

Reply:
Assets should now use `Asset` class. 

Docs could be updated to recommend use `Router::url()` for MVC type actions first, also keep support what currently checked with `$plainString`, other bare string uses should avoided. `Router::url()` is now used basically for any type of URL, should be less usages.

`Asset` itself should not use `Router::url()` - it currently uses it with scheme URLs, but URL is just returned unmodified via `$plainString` check. And the `a-z:` part doesn't not work when base URL is used: `foo:bar` becomes `base/foo:bar` (except `mailto/javascript/tel/sms`) - I'm not sure if this behavior is intended.